omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.providers.r3;
+
+import javax.validation.constraints.NotNull;
+
+import java.util.List;
+
+import ca.uhn.fhir.rest.annotation.*;
+import ca.uhn.fhir.rest.api.SortSpec;
+import ca.uhn.fhir.rest.api.server.IBundleProvider;
+import ca.uhn.fhir.rest.param.*;
+import ca.uhn.fhir.rest.server.IResourceProvider;
+import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
+import lombok.AccessLevel;
+import lombok.Setter;
+import org.hl7.fhir.convertors.VersionConvertor_30_40;
+import org.hl7.fhir.dstu3.model.*;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.openmrs.module.fhir2.api.FhirObservationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+@Qualifier("fhirR3Resources")
+@Setter(AccessLevel.PACKAGE)
+public class ObservationFhirResourceProvider implements IResourceProvider {
+	
+	@Autowired
+	private FhirObservationService observationService;
+	
+	@Override
+	public Class<? extends IBaseResource> getResourceType() {
+		return Observation.class;
+	}
+	
+	@Read
+	@SuppressWarnings("unused")
+	public Observation getObservationById(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Observation observation = observationService.get(id.getIdPart());
+		if (observation == null) {
+			throw new ResourceNotFoundException("Could not find observation with Id " + id.getIdPart());
+		}
+		
+		return VersionConvertor_30_40.convertObservation(observation);
+	}
+	
+	@History
+	public List<Resource> getObservationHistoryById(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Observation observation = observationService.get(id.getIdPart());
+		if (observation == null) {
+			throw new ResourceNotFoundException("Could not find Observation with Id " + id.getIdPart());
+		}
+		return VersionConvertor_30_40.convertObservation(observation).getContained();
+	}
+	
+	@Search
+	public IBundleProvider searchObservations(
+	        @OptionalParam(name = Observation.SP_ENCOUNTER) ReferenceAndListParam encounterReference,
+	        @OptionalParam(name = Observation.SP_SUBJECT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_GIVEN,
+	                Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam patientReference,
+	        //TODO: SP_HAS_MEMBER not present so converted to SP_BASED_ON just to prevent errors
```

Comment:
This should be `SP_RELATED`, but note that `SP_RELATED` has a somewhat different structure.

Reply:
@ibacher Shouldn't this be `SP_RELATED_TYPE` [here](http://hl7.org/fhir/STU3/observation.html#search) is why I think so.
Also only changing the data type will not solve the problems, I think that we'll have to map it, as we did for `Task` also the `observation` resource returns `IBundleProvider` in search function which complicates things or if not all the tests are failing with this error message
```
ca.uhn.fhir.context.ConfigurationException: This context is for FHIR version "DSTU3" but the class "org.hl7.fhir.r4.model.OperationOutcome" is for version "R4"

```
I guess we'll need to override this method also in `BaseFhirR3ResourceProviderTest`

Reply:
No. `SP_RELATED`. `SP_RELATED_TYPE` refers to the type structure that is part of the related element. Technically, `SP_RELATED_TARGET` is the more direct replacement for `SP_HAS_MEMBER`, but only when the `related` structure looks like this:

```json
"related": {
    "type": "has-member",
    "target": "Observation/3"
}
```

Or, in terms of search parameters:

```
/Observation?related.type=has-member&related.target=Observation/3
```

This also isn't exactly like Task, because what we're converting here isn't a full resource, but a search parameter.

The message you highlight:

```
ca.uhn.fhir.context.ConfigurationException: This context is for FHIR version "DSTU3" but the class "org.hl7.fhir.r4.model.OperationOutcome" is for version "R4"
```

Does indicate a problem, but part of the problem is that we're not properly handling things when the result is an `OperationOutcome` (which itself needs to be translated). Please step through one of these failing tests with a debugger to see what is actually happening.

> I guess we'll need to override this method also in BaseFhirR3ResourceProviderTest

Which method are you referring to?

Reply:
> Which method are you referring to?


protected void describeMismatchSafely(MockHttpServletResponse item, Description mismatchDescription) in this function we have `OperationOutcome` which is of `import org.hl7.fhir.r4.model.OperationOutcome;` type so I think this needs to be generic so that all versions and use it